File: tg-bot.js
const fs = require('fs');
const path = require('path');
const TelegramBot = require('node-telegram-bot-api');

const TOKEN = process.env.TG_BOT_TOKEN;
if (!TOKEN) {
  throw new Error('Set TG_BOT_TOKEN in environment variables');
}

const ADMIN_CHAT_ID = process.env.ADMIN_CHAT_ID || null;
const DB_PATH = path.join(__dirname, 'products.json');

const bot = new TelegramBot(TOKEN, { polling: true });

function readProducts() {
  if (!fs.existsSync(DB_PATH)) {
    fs.writeFileSync(DB_PATH, JSON.stringify([], null, 2));
  }
  const raw = fs.readFileSync(DB_PATH, 'utf8');
  const parsed = JSON.parse(raw);
  return Array.isArray(parsed) ? parsed : [];
}

function writeProducts(products) {
  fs.writeFileSync(DB_PATH, JSON.stringify(products, null, 2));
}

function formatProducts(products) {
  if (!products.length) return 'Список товарів порожній.';
  return products
    .map(
      p =>
        `ID: ${p.id}\nНазва: ${p.name}\nЦіна: ${p.price} грн\nКількість: ${p.qty}`
    )
    .join('\n\n');
}

function parseArgs(text) {
  return text.trim().split(/\s+/);
}

bot.onText(/^\/start$/, msg => {
  bot.sendMessage(
    msg.chat.id,
    [
      'Привіт! Це бот для керування замовленнями 🛒',
      '',
      'Команди:',
      '/list - список товарів',
      '/add <id> <назва> <ціна> <кількість> - додати товар',
      '/setqty <id> <кількість> - змінити кількість',
      '/order <id> <кількість> <імʼя> <номер> - оформити замовлення',
      '/remove <id> - видалити товар'
    ].join('\n')
  );
});

bot.onText(/^\/list$/, msg => {
  const products = readProducts();
  bot.sendMessage(msg.chat.id, `📦 Товари:\n\n${formatProducts(products)}`);
});

bot.onText(/^\/add\s+(.+)/, (msg, match) => {
  const args = parseArgs(match[1]);
  if (args.length < 4) {
    bot.sendMessage(msg.chat.id, 'Формат: /add <id> <назва> <ціна> <кількість>');
    return;
  }

  const [id, ...rest] = args;
  const qty = Number(rest.pop());
  const price = Number(rest.pop());
  const name = rest.join(' ');

  if (!name || Number.isNaN(price) || Number.isNaN(qty)) {
    bot.sendMessage(msg.chat.id, 'Некоректні дані.');
    return;
  }

  const products = readProducts();
  const exists = products.find(p => p.id === id);
  if (exists) {
    bot.sendMessage(msg.chat.id, `Товар з ID ${id} вже існує.`);
    return;
  }

  products.push({ id, name, price, qty });
  writeProducts(products);
  bot.sendMessage(msg.chat.id, `✅ Додано товар:\nID: ${id}\nНазва: ${name}\nКількість: ${qty}`);
});

bot.onText(/^\/setqty\s+(.+)/, (msg, match) => {
  const [id, qtyText] = parseArgs(match[1]);
  const qty = Number(qtyText);
  if (!id || Number.isNaN(qty) || qty < 0) {
    bot.sendMessage(msg.chat.id, 'Формат: /setqty <id> <кількість>');
    return;
  }

  const products = readProducts();
  const product = products.find(p => p.id === id);
  if (!product) {
    bot.sendMessage(msg.chat.id, `Товар з ID ${id} не знайдено.`);
    return;
  }

  product.qty = qty;
  writeProducts(products);
  bot.sendMessage(msg.chat.id, `✅ Кількість оновлено. ${product.name}: ${qty}`);
});

bot.onText(/^\/remove\s+(.+)/, (msg, match) => {
  const [id] = parseArgs(match[1]);
  const products = readProducts();
  const next = products.filter(p => p.id !== id);

  if (next.length === products.length) {
    bot.sendMessage(msg.chat.id, `Товар з ID ${id} не знайдено.`);
    return;
  }

  writeProducts(next);
  bot.sendMessage(msg.chat.id, `🗑️ Товар з ID ${id} видалено.`);
});

bot.onText(/^\/order\s+(.+)/, (msg, match) => {
  const args = parseArgs(match[1]);
  if (args.length < 4) {
    bot.sendMessage(msg.chat.id, 'Формат: /order <id> <кількість> <імʼя> <номер>');
    return;
  }

  const id = args[0];
  const qty = Number(args[1]);
  const phone = args[args.length - 1];
  const customerName = args.slice(2, -1).join(' ');

  if (!customerName || Number.isNaN(qty) || qty <= 0) {
    bot.sendMessage(msg.chat.id, 'Некоректні дані замовлення.');
    return;
  }

  const products = readProducts();
  const product = products.find(p => p.id === id);

  if (!product) {
    bot.sendMessage(msg.chat.id, `Товар з ID ${id} не знайдено.`);
    return;
  }

  if (product.qty < qty) {
    bot.sendMessage(msg.chat.id, `Недостатньо товару. В наявності: ${product.qty}`);
    return;
  }

  product.qty -= qty;
  writeProducts(products);

  const orderText = [
    '🛍️ НОВЕ ЗАМОВЛЕННЯ',
    '',
    `ID товару: ${product.id}`,
    `Назва: ${product.name}`,
    `Кількість: ${qty}`,
    `Сума: ${qty * product.price} грн`,
    '',
    '👤 Дані клієнта:',
    `Імʼя: ${customerName}`,
    `Номер: ${phone}`,
    '',
    `Залишок на складі: ${product.qty}`
  ].join('\n');

  bot.sendMessage(msg.chat.id, `✅ Замовлення оформлено!\n\n${orderText}`);

  if (ADMIN_CHAT_ID) {
    bot.sendMessage(ADMIN_CHAT_ID, orderText);
  }
});

console.log('Telegram bot is running...');

File: tg_bot.py
import json
import os
from pathlib import Path
from telegram import Update
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes

DB_PATH = Path(__file__).with_name('products.json')


def read_products():
    if not DB_PATH.exists():
        DB_PATH.write_text('[]', encoding='utf-8')
    data = json.loads(DB_PATH.read_text(encoding='utf-8'))
    return data if isinstance(data, list) else []


def write_products(products):
    DB_PATH.write_text(json.dumps(products, ensure_ascii=False, indent=2), encoding='utf-8')


def format_products(products):
    if not products:
        return 'Список товарів порожній.'
    return '\n\n'.join(
        f"ID: {p['id']}\nНазва: {p['name']}\nЦіна: {p['price']} грн\nКількість: {p['qty']}"
        for p in products
    )


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    text = (
        'Привіт! Python-бот для замовлень 🛒\n\n'
        'Команди:\n'
        '/list - список товарів\n'
        '/add <id> <ціна> <кількість> <назва>\n'
        '/setqty <id> <кількість>\n'
        '/remove <id>\n'
        '/order <id> <кількість> <імʼя> <номер>'
    )
    await update.message.reply_text(text)


async def list_products(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(f"📦 Товари:\n\n{format_products(read_products())}")


async def add_product(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if len(context.args) < 4:
        await update.message.reply_text('Формат: /add <id> <ціна> <кількість> <назва>')
        return
    pid = context.args[0]
    try:
        price = float(context.args[1])
        qty = int(context.args[2])
    except ValueError:
        await update.message.reply_text('Ціна або кількість некоректні.')
        return
    name = ' '.join(context.args[3:]).strip()
    products = read_products()
    if any(p['id'] == pid for p in products):
        await update.message.reply_text(f'Товар з ID {pid} вже існує.')
        return
    products.append({'id': pid, 'name': name, 'price': price, 'qty': qty})
    write_products(products)
    await update.message.reply_text(f'✅ Додано: {name} (ID: {pid})')


async def set_qty(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if len(context.args) != 2:
        await update.message.reply_text('Формат: /setqty <id> <кількість>')
        return
    pid, qty_raw = context.args
    try:
        qty = int(qty_raw)
    except ValueError:
        await update.message.reply_text('Кількість має бути числом.')
        return
    products = read_products()
    product = next((p for p in products if p['id'] == pid), None)
    if not product:
        await update.message.reply_text(f'Товар з ID {pid} не знайдено.')
        return
    product['qty'] = qty
    write_products(products)
    await update.message.reply_text(f'✅ Оновлено залишок: {product["name"]} = {qty}')


async def remove_product(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if len(context.args) != 1:
        await update.message.reply_text('Формат: /remove <id>')
        return
    pid = context.args[0]
    products = read_products()
    filtered = [p for p in products if p['id'] != pid]
    if len(filtered) == len(products):
        await update.message.reply_text(f'Товар з ID {pid} не знайдено.')
        return
    write_products(filtered)
    await update.message.reply_text(f'🗑️ Товар {pid} видалено.')


async def order_product(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if len(context.args) < 4:
        await update.message.reply_text('Формат: /order <id> <кількість> <імʼя> <номер>')
        return
    pid = context.args[0]
    try:
        qty = int(context.args[1])
    except ValueError:
        await update.message.reply_text('Кількість має бути числом.')
        return
    phone = context.args[-1]
    customer_name = ' '.join(context.args[2:-1]).strip()
    products = read_products()
    product = next((p for p in products if p['id'] == pid), None)
    if not product:
        await update.message.reply_text(f'Товар з ID {pid} не знайдено.')
        return
    if product['qty'] < qty:
        await update.message.reply_text(f'Недостатньо товару. В наявності: {product["qty"]}')
        return
    product['qty'] -= qty
    write_products(products)
    total = qty * product['price']
    text = (
        '🛍️ НОВЕ ЗАМОВЛЕННЯ\n\n'
        f'ID товару: {product["id"]}\n'
        f'Назва: {product["name"]}\n'
        f'Кількість: {qty}\n'
        f'Сума: {total} грн\n\n'
        '👤 Дані клієнта:\n'
        f'Імʼя: {customer_name}\n'
        f'Номер: {phone}\n\n'
        f'Залишок: {product["qty"]}'
    )
    await update.message.reply_text(text)


def main():
    token = os.getenv('TG_BOT_TOKEN')
    if not token:
        raise RuntimeError('Set TG_BOT_TOKEN env variable')
    app = ApplicationBuilder().token(token).build()
    app.add_handler(CommandHandler('start', start))
    app.add_handler(CommandHandler('list', list_products))
    app.add_handler(CommandHandler('add', add_product))
    app.add_handler(CommandHandler('setqty', set_qty))
    app.add_handler(CommandHandler('remove', remove_product))
    app.add_handler(CommandHandler('order', order_product))
    app.run_polling()


if __name__ == '__main__':
    main()
